src/main/java/org/fcrepo/storage/ocfl/DefaultOcflObjectSession.java

```diff
@@ -0,0 +1,477 @@
+
+        if (Files.exists(stagingPath)) {
+            try {
+                return Optional.of(Files.newInputStream(stagingPath));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<InputStream> readFromOcfl(final PathPair path) {
+        if (ocflRepo.containsObject(ocflObjectId)) {
+            final var object = ocflRepo.getObject(ObjectVersionId.head(ocflObjectId));
+            if (object.containsFile(path.path)) {
+                return Optional.of(object.getFile(path.path).getStream());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Path stagingPath(final PathPair path) {
+        return objectStaging.resolve(path.encodedPath);
+    }
+
+    private Path createStagingPath(final PathPair path) {
+        final var stagingPath = stagingPath(path);
+
+        try {
+            Files.createDirectories(stagingPath.getParent());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return stagingPath;
+    }
+
+    private void write(final Optional<InputStream> content, final Path destination) {
+        if (content.isPresent()) {
+            try {
+                Files.copy(content.get(), destination, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    private void writeHeaders(final ResourceHeaders headers, final Path destination) {
+        try {
+            objectMapper.writeValue(destination.toFile(), headers);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void deletePathsFromStaging() {
+        deletePaths.stream().map(this::stagingPath).forEach(path -> {
+            if (Files.exists(path)) {
+                try {
+                    Files.delete(path);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+        });
+    }
+
+    private ResourceHeaders readHeaders(final InputStream stream) {
+        try {
+            return objectMapper.readValue(stream, ResourceHeaders.class);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean newInSession(final PathPair headerPath) {
+        if (ocflRepo.containsObject(ocflObjectId)) {
+            return !ocflRepo.describeVersion(ObjectVersionId.head(ocflObjectId)).containsFile(headerPath.path);
+        }
+        return true;
+    }
+
+    private String loadRootResourceId() {
+        final var stream = readFromOcfl(encode(PersistencePaths.ROOT_HEADER_PATH));
+
+        if (stream.isPresent()) {
+            final var headers = readHeaders(stream.get());
+            return headers.getId();
+        }
+
+        return null;
```

Comment:
What does it mean if we can't load a root resource? Can this state be recovered from?

Reply:
It should mean that the OCFL object does not exist, as is the case when you're creating a new resource. In this case, the root resource id for the session is set on the first `write` call.

It a pathological case, it could mean that the OCFL object exists, but does not contain a valid F6 resource. In this case, the behavior is undefined.

Reply:
Added javadocs, and changed it to throw an exception if the object exists but a root resource could not be found.